atlasdb-commons/src/main/java/com/palantir/util/ExceptionHandlingRunner.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+
+/**
+ * Runs runnables and suppliers that may throw exceptions, and swallows those exceptions until later. Can be used as a
+ * resource in a try block, or delegated to by another resource.
+ */
+public final class ExceptionHandlingRunner implements AutoCloseable {
+    private final List<Throwable> failures = new ArrayList<>();
+
+    public ExceptionHandlingRunner() {}
+
+    /**
+     * Instantiates the runner with a throwable that has already been caught, to be rethrown when close is complete.
+     */
+    public ExceptionHandlingRunner(Throwable t) {
+        failures.add(t);
+    }
+
+    public void runSafely(Runnable shutdownCallback) {
+        try {
+            shutdownCallback.run();
+        } catch (Throwable throwable) {
+            failures.add(throwable);
+        }
+    }
+
+    public <T> Optional<T> supplySafely(Supplier<T> shutdownCallback) {
+        try {
+            return Optional.of(shutdownCallback.get());
+        } catch (Throwable throwable) {
+            failures.add(throwable);
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Calling close with no failures should be a no-op; equally, calling close multiple times will re-throw a runtime
+     * exception with the same suppressed errors (plus any additional errors suppressed since the last close call).
```

Comment:
OK. Do we want to make any guarantees on the ordering? Might be good to add a test to ratify whichever version we want (in the order executed, or just that it contains the same elements - I imagine the assertion picked is based on the guarantee we decide here).

Reply:
Intended behaviour is that they are added to the suppressed list in the order of being called - please may you verify that my test does indeed confirm this behaviour? Should I add this to the javadoc?

Reply:
Yeah, the test does confirm that.

This class is in `atlasdb-commons` so I'd bias towards being very explicit with the API, someone in large internal product might actually pick it up and use it outside of Atlas code.